convex/engine/game.ts

```diff
@@ -128,37 +128,59 @@ export abstract class Game<Handlers extends InputHandlers> {
       currentTs = candidateTs;
     }
 
-    let nextRun = this.idleUntil(currentTs);
+    let stepNextRun = this.idleUntil(currentTs);
 
     // Force an immediate wakeup if we have more inputs to process or more time to simulate.
     if (inputs.length === this.maxInputsPerStep) {
       console.warn(`Received max inputs (${this.maxInputsPerStep}) for step`);
-      nextRun = null;
+      stepNextRun = null;
     }
     if (numTicks === this.maxTicksPerStep) {
       console.warn(`Only simulating ${currentTs - startTs}ms due to max ticks per step limit.`);
-      nextRun = null;
+      stepNextRun = null;
     }
-    nextRun = nextRun ?? now + this.stepDuration;
+    stepNextRun = stepNextRun ?? now + this.stepDuration;
 
     // Commit the step by moving time forward, consuming our inputs, and saving the game's state.
     await this.save(ctx);
-    const nextGenerationNumber = generationNumber + 1;
     await ctx.db.patch(engine._id, {
       currentTime: currentTs,
       lastStepTs,
       processedInputNumber,
-      state: { kind: 'running', nextRun },
-      generationNumber: nextGenerationNumber,
-    });
-    await ctx.scheduler.runAt(nextRun, stepReference, {
-      engineId: this.engineId,
-      generationNumber: nextGenerationNumber,
+      // Use the generation number to serialize all instances of `runStep`.
+      generationNumber: engine.generationNumber + 1,
```

Comment:
this is just a safeguard in case no other writes conflicted during the step? From what I can tell, nextGenerationNumber should have also serialized execution previously - is that right?

Reply:
yep, exactly! the main difference w/this algorithm is that we're not sending the generation number as an argument to the scheduled jobs in the future (to be able to cancel them by bumping the number). we're just using it for forcing a write conflict if no other writes conflict.